### PR TITLE
[New Record] Simplified HC into single cached activation. -about 0.5s

### PR DESCRIPTION
**Pull request by ClassicLarry**
Dropping roughly half a second (ran on a slower machine than the timing used on the record right now), I'll get an accurate retiming when I go to merge.
Also bringing the loss much more comfortably under 3.28.

The partitioned hyperconnections PR indicated that the last few attention layers strongly prefer to ignore the contributions of the last few MLPs. I made this explicit by instead of doing hyperconnections, feeding in a single cached activation from layer 7 for all later attention modules, so they never get polluted with 'prediction MLPs'. 

The block loop then compresses into something quite simple:
```
x_backout = None
for i in range(num_layers):
    attn_in = x_backout if x_backout is not None else x 
    x = x + attn(norm(attn_in))
    x = x + mlp(norm(x))
    # last 3 attn layers will query saved context vector
    # the remaining residual stream builds predictions without polluting attn input
    if i == num_layers-4: 
        x_backout = x
# reverse out context, leaving only prediction
x = x - backout_lambda*x_backout 
```

I left the PR at this stage because it opens up two valid follow ons: 1) now that 3 attention layers all have the exact same input, consider merging them into 1 mega layer with shared matmuls. OR 2) don't have all 3 attention layers accept the exact same input, enable some sequential evolution of this input.

```
import scipy.stats
import torch
losses = [3.2784, 3.2791, 3.277, 3.2766, 3.2787, 3.2778, 3.279, 3.2776]
times = [87.098, 87.225, 87.276, 87.226, 87.214, 87.263, 87.316, 87.222]
print("p=%.4f" % scipy.stats.ttest_1samp(losses, 3.28, alternative="less").pvalue)
# p=0.0003
print("losses:", torch.std_mean(torch.tensor(losses)))
# losses: (tensor(0.0009), tensor(3.2780))
print("time:", torch.std_mean(torch.tensor(times)))
# time: (tensor(0.0637), tensor(87.2300))
```

### CODE REVIEW

**Comment by ClassicLarry** [comment]
Plan to merge later at -0.7s at 86.1s. Hopefully someone will be able to improve on this one- at a minimum the norm(x_backout) is computed 3 times. the last 3 layers have (paired head short window, short window, long window). So, may have some interesting design decisions on how to construct a mega attn block.

**Comment by ClassicLarry** [comment]
When I went to retime this the first run came back with 3.281 loss. So, I let it run for 9 more iterations to confirm there wasn't something weird going on with interaction effects of last few PRs. [3.2810, 3.2784, 3.2801, 3.2785, 3.2778, 3.2807, 3.2788, 3.2776, 3.2780, 3.2771]. Looks fine.

**Comment by QueenFi703** [comment]
This approach is *more* promising than mega-layer merging for several
reasons:

*Why Sequential Evolution Wins:*

   1. Maintains independent learned projections (QKVO for each layer)
   2. Allows subtle context drift through the 3-layer sequence
   3. Empirically, the 3-layer window may benefit from *slight* variation,
   not uniformity
   4. Matches your observation: "last 3 layers have (paired head short
   window, short window, long window)"
   5. Incremental cost vs. exponential benefit






# ---- Transformer layers ----
x_backout = None
skip_connection = None
for i in range(self.num_layers):
yarn = self.yarn_paired_head if i in self.paired_head_layers else self.yarn
attn_args = AttnArgs(...)

qkvo_w = self.attn_qkvo_weights[i]
attn = self.attn_paired if i in self.paired_head_layers else self.attn
mlp_fcs = self.mlp_fc_weights
mlp_projs = self.mlp_proj_weights
c_fc = mlp_fcs[i]
c_proj = mlp_projs[i]

# Special handling for layer 6 (no attention, inject skip)
if i == 6:
x = x + skip_gate_out * skip_connection
else:
# Determine attention input source
if i >= 8 and x_backout is not None:
# For layers 8+, evolve from cached activation progressively
attn_in = x_backout + (x - x_backout) * (i - 8) * 0.5 # 0.0, 0.5, 1.0 blend
else:
attn_in = x

# Precompute norm once and reuse
attn_in_norm = norm(attn_in)
attn_out = attn(attn_in_norm, attn_args, qkvo_w)
x = resid_lambdas_attn[i] * x + post_lambdas_attn[i] * attn_out +
x0_inject[i]

# MLP on evolved state (always on full x)
x = resid_lambdas_mlp[i] * x + post_lambdas_mlp[i] * ReLUSqrdMLP(norm(x),
c_fc, c_proj)

# Cache key activations
if i == 3:
skip_connection = x
if i == 7:
x_backout = x.detach() # Detach to save gradients

Issue Current Proposed
Redundant norm() calls 3x on same input 0x (blend eliminates redundancy)
Context evolution Static (identical 3x) Gradual (0.0→1.0 interpolation)
Memory efficiency 3 full param sets Same 3, but strategic blending
Gradient flow Static reference Dynamic, learns blend schedule
Computational gain ~0.5s ~0.6

**Comment by kroggen** [comment]
This is really cool!

I perceived that the last 3 layers update `x` without paying attention to the content that is already on `x`, so I thought that the next possible step would be to make a modification that includes `x` on the attention part, to make it conditional on it as well

I used cross-attention where the queries come from `x` and the keys and values come from `x_backout`, but unhappily the validation loss is worse (3.2820 if I remember correctly)

Diff here:

```diff
diff --git a/train_gpt.py b/train_gpt.py
index b3d753b..622833d 100644
--- a/train_gpt.py
+++ b/train_gpt.py
@@ -1072,7 +1072,15 @@ class CausalSelfAttention(nn.Module):
         assert self.hdim == self.dim, "num_heads * head_dim must equal model_dim"
         # Weights are stored in parameter banks and passed via forward()
 
-    def forward(self, x: Tensor, attn_args: AttnArgs, qkvo_w: Tensor):
+    def forward(self, x: Tensor, attn_args: AttnArgs, qkvo_w: Tensor, q_x: Tensor = None):
+        """
+        Args:
+            x: Input for K/V (and Q if q_x is None). Shape: (B, T, dim)
+            attn_args: Attention configuration
+            qkvo_w: Combined Q, K, V, O weights
+            q_x: Optional separate input for Q (cross-attention mode). Shape: (B, T, dim)
+                 If provided, Q comes from q_x, K/V come from x.
+        """
         B, T = x.size(0), x.size(1) # batch size, sequence length
         assert B == 1, "varlen sequences requires B == 1"
         assert T % 16 == 0
@@ -1085,11 +1093,18 @@ class CausalSelfAttention(nn.Module):
         attn_gate_w, ve_gate_w = attn_args.attn_gate_w, attn_args.ve_gate_w
         train_max_seq_len = attn_args.train_max_seq_len
 
-        q, k, v = F.linear(x, sa_lambdas[0] * qkvo_w[:self.dim * 3].type_as(x)).view(B, T, 3 * self.num_heads, self.head_dim).chunk(3, dim=-2)
+        cross_attn = q_x is not None
+        if cross_attn:
+            q = F.linear(q_x, sa_lambdas[0] * qkvo_w[:self.dim].type_as(q_x)).view(B, T, self.num_heads, self.head_dim)
+            k, v = F.linear(x, sa_lambdas[0] * qkvo_w[self.dim:self.dim * 3].type_as(x)).view(B, T, 2 * self.num_heads, self.head_dim).chunk(2, dim=-2)
+        else:
+            q, k, v = F.linear(x, sa_lambdas[0] * qkvo_w[:self.dim * 3].type_as(x)).view(B, T, 3 * self.num_heads, self.head_dim).chunk(3, dim=-2)
         max_len = train_max_seq_len if self.training else (args.val_batch_size // (grad_accum_steps * world_size))
 
         q, k = norm(q), norm(k) # QK norm @Grad62304977
 
+        x_for_gates = q_x if cross_attn else x
+
         if not self.paired:
             q, k = yarn.rotary(q), yarn.rotary(k)
 
@@ -1099,7 +1114,7 @@ class CausalSelfAttention(nn.Module):
 
             if ve is not None:
                 # gate pattern g(x[:6] + ve[:6]) by @photomz
-                ve_gate_out = 2 * torch.sigmoid(F.linear(torch.cat([x[..., :6], ve[None, ..., :6]], dim=-1), ve_gate_w)).view(B, T, self.num_heads, 1)
+                ve_gate_out = 2 * torch.sigmoid(F.linear(torch.cat([x_for_gates[..., :6], ve[None, ..., :6]], dim=-1), ve_gate_w)).view(B, T, self.num_heads, 1)
                 v = v + ve_gate_out * ve.view_as(v) # @ KoszarskyB & @Grad62304977
 
         else:
@@ -1117,7 +1132,7 @@ class CausalSelfAttention(nn.Module):
             k = k.view(B, T * 2, self.num_heads // 2, self.head_dim)
 
             if ve is not None:
-                ve_gate_out = 2 * torch.sigmoid(F.linear(x[..., :12], ve_gate_w)).view(B, T * 2, self.num_heads // 2, 1)
+                ve_gate_out = 2 * torch.sigmoid(F.linear(x_for_gates[..., :12], ve_gate_w)).view(B, T * 2, self.num_heads // 2, 1)
                 v = v + ve_gate_out * ve.view_as(v)
 
             seqlens = 2 * seqlens
@@ -1128,7 +1143,7 @@ class CausalSelfAttention(nn.Module):
                                                         max_seqlen_q=max_len, max_seqlen_k=max_len,
                                                         causal=True, softmax_scale=yarn.attn_scale, window_size=(bm_size, 0))
         y = y.view(B, T, self.num_heads, self.head_dim)
-        y = y * torch.sigmoid(F.linear(x[..., :12], attn_gate_w)).view(B, T, self.num_heads, 1)
+        y = y * torch.sigmoid(F.linear(x_for_gates[..., :12], attn_gate_w)).view(B, T, self.num_heads, 1)
         y = y.contiguous().view(B, T, self.num_heads * self.head_dim) # re-assemble all head outputs side by side
         y = F.linear(y, sa_lambdas[1] * qkvo_w[self.dim * 3:].type_as(y))  # sa_lambdas[1] pre-multiplied to O @shenberg
         return y
@@ -1333,8 +1348,10 @@ class GPT(nn.Module):
             if i == 6:
                 x = x + skip_gate_out * skip_connection
             else:
-                attn_in = x_backout if x_backout is not None else x
-                attn_out = attn(norm(attn_in), attn_args, qkvo_w)
+                if x_backout is not None:
+                    attn_out = attn(norm(x_backout), attn_args, qkvo_w, q_x=norm(x))
+                else:
+                    attn_out = attn(norm(x), attn_args, qkvo_w)
                 x = resid_lambdas_attn[i] * x + post_lambdas_attn[i] * attn_out + x0_inject[i]
             x = resid_lambdas_mlp[i] * x + post_lambdas_mlp[i] * ReLUSqrdMLP(norm(x), c_fc, c_proj)
             if i == 3:
```

On layer 8 the output is the same (`x` and `x_backout` are the same) so it is really cross-attention only on the last 2 layers

Layer 9 uses paired-head attention and layer 10 uses the long-window path with key_offset, so not sure if they can conflict with the cross-attention

**Comment by ClassicLarry** [comment]
> This is really cool!
> 
> I perceived that the last 3 layers update `x` without paying attention to the content that is already on `x`, so I thought that the next possible step would be to make a modification that includes `x` on the attention part, to make it conditional on it as well
> 
> I used cross-attention where the queries come from `x` and the keys and values come from `x_backout`, but unhappily the validation loss is worse (3.2820 if I remember correctly)
> 
> Diff here:
> 
> ```diff
> diff --git a/train_gpt.py b/train_gpt.py
> index b3d753b..622833d 100644
> --- a/train_gpt.py
> +++ b/train_gpt.py
> @@ -1072,7 +1072,15 @@ class CausalSelfAttention(nn.Module):
>          assert self.hdim == self.dim, "num_heads * head_dim must equal model_dim"
>          # Weights are stored in parameter banks and passed via forward()
>  
> -    def forward(self, x: Tensor, attn_args: AttnArgs, qkvo_w: Tensor):
> +    def forward(self, x: Tensor, attn_args: AttnArgs, qkvo_w: Tensor, q_x: Tensor = None):
> +        """
> +        Args:
> +            x: Input for K/V (and Q if q_x is None). Shape: (B, T, dim)
> +            attn_args: Attention configuration
> +            qkvo_w: Combined Q, K, V, O weights
> +            q_x: Optional separate input for Q (cross-attention mode). Shape: (B, T, dim)
> +                 If provided, Q comes from q_x, K/V come from x.
> +        """
>          B, T = x.size(0), x.size(1) # batch size, sequence length
>          assert B == 1, "varlen sequences requires B == 1"
>          assert T % 16 == 0
> @@ -1085,11 +1093,18 @@ class CausalSelfAttention(nn.Module):
>          attn_gate_w, ve_gate_w = attn_args.attn_gate_w, attn_args.ve_gate_w
>          train_max_seq_len = attn_args.train_max_seq_len
>  
> -        q, k, v = F.linear(x, sa_lambdas[0] * qkvo_w[:self.dim * 3].type_as(x)).view(B, T, 3 * self.num_heads, self.head_dim).chunk(3, dim=-2)
> +        cross_attn = q_x is not None
> +        if cross_attn:
> +            q = F.linear(q_x, sa_lambdas[0] * qkvo_w[:self.dim].type_as(q_x)).view(B, T, self.num_heads, self.head_dim)
> +            k, v = F.linear(x, sa_lambdas[0] * qkvo_w[self.dim:self.dim * 3].type_as(x)).view(B, T, 2 * self.num_heads, self.head_dim).chunk(2, dim=-2)
> +        else:
> +            q, k, v = F.linear(x, sa_lambdas[0] * qkvo_w[:self.dim * 3].type_as(x)).view(B, T, 3 * self.num_heads, self.head_dim).chunk(3, dim=-2)
>          max_len = train_max_seq_len if self.training else (args.val_batch_size // (grad_accum_steps * world_size))
>  
>          q, k = norm(q), norm(k) # QK norm @Grad62304977
>  
> +        x_for_gates = q_x if cross_attn else x
> +
>          if not self.paired:
>              q, k = yarn.rotary(q), yarn.rotary(k)
>  
> @@ -1099,7 +1114,7 @@ class CausalSelfAttention(nn.Module):
>  
>              if ve is not None:
>                  # gate pattern g(x[:6] + ve[:6]) by @photomz
> -                ve_gate_out = 2 * torch.sigmoid(F.linear(torch.cat([x[..., :6], ve[None, ..., :6]], dim=-1), ve_gate_w)).view(B, T, self.num_heads, 1)
> +                ve_gate_out = 2 * torch.sigmoid(F.linear(torch.cat([x_for_gates[..., :6], ve[None, ..., :6]], dim=-1), ve_gate_w)).view(B, T, self.num_heads, 1)
>                  v = v + ve_gate_out * ve.view_as(v) # @ KoszarskyB & @Grad62304977
>  
>          else:
> @@ -1117,7 +1132,7 @@ class CausalSelfAttention(nn.Module):
>              k = k.view(B, T * 2, self.num_heads // 2, self.head_dim)
>  
>              if ve is not None:
> -                ve_gate_out = 2 * torch.sigmoid(F.linear(x[..., :12], ve_gate_w)).view(B, T * 2, self.num_heads // 2, 1)
> +                ve_gate_out = 2 * torch.sigmoid(F.linear(x_for_gates[..., :12], ve_gate_w)).view(B, T * 2, self.num_heads // 2, 1)
>                  v = v + ve_gate_out * ve.view_as(v)
>  
>              seqlens = 2 * seqlens
> @@ -1128,7 +1143,7 @@ class CausalSelfAttention(nn.Module):
>                                                          max_seqlen_q=max_len, max_seqlen_k=max_len,
>                                                          causal=True, softmax_scale=yarn.attn_scale, window_size=(bm_size, 0))
>          y = y.view(B, T, self.num_heads, self.head_dim)
> -        y = y * torch.sigmoid(F.linear(x[..., :12], attn_gate_w)).view(B, T, self.num_heads, 1)
> +        y = y * torch.sigmoid(F.linear(x_for_gates[..., :12], attn_gate_w)).view(B, T, self.num_heads, 1)
>          y = y.contiguous().view(B, T, self.num_heads * self.head_dim) # re-assemble all head outputs side by side
>          y = F.linear(y, sa_lambdas[1] * qkvo_w[self.dim * 3:].type_as(y))  # sa_lambdas[1] pre-multiplied to O @shenberg
>          return y
> @@ -1333,8 +1348,10 @@ class GPT(nn.Module):
>              if i == 6:
>                  x = x + skip_gate_out * skip_connection
>              else:
> -                attn_in = x_backout if x_backout is not None else x
> -                attn_out = attn(norm(attn_in), attn_args, qkvo_w)
> +                if x_backout is not None:
> +                    attn_out = attn(norm(x_backout), attn_args, qkvo_w, q_x=norm(x))
> +                else:
> +                    attn_out = attn(norm(x), attn_args, qkvo_w)
>                  x = resid_lambdas_attn[i] * x + post_lambdas_attn[i] * attn_out + x0_inject[i]
>              x = resid_lambdas_mlp[i] * x + post_lambdas_mlp[i] * ReLUSqrdMLP(norm(x), c_fc, c_proj)
>              if i == 3:
> ```
> 
> On layer 8 the output is the same (`x` and `x_backout` are the same) so it is really cross-attention only on the last 2 layers
> 
> Layer 9 uses paired-head attention and layer 10 uses the long-window path with key_offset, so not sure if they can conflict with the cross-attention

I like this idea, nice intuition. For testing purposes, it might be worth temporarily dropping paired head attn from that layer and comparing to that baseline to look for signs of life. You can also do things like let the query be lambda1*xbackout+lambda2*x. Then see what the lambdas become. That way you let the model tell you what it’s trying to do. Same for keys and values. Anywhere the lambdas learn some non uniform behavior, there is a signal.
